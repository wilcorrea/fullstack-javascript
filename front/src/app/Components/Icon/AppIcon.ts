import { CreateElement, RenderContext } from 'vue'

/**
 * @type {string}
 */
const block = 'app-icon'

/**
 * @component {AppIcon}
 */
export default {
  /**
   */
  functional: true,
  /**
   */
  name: 'AppIcon',
  /**
   * @param {CreateElement} h
   * @param {RenderContext} context
   */
  render (h: CreateElement, context: RenderContext) {
    const { props, listeners } = context
    const style = {}
    if (props.color) {
      // @ts-ignore
      style['color'] = props.color
    }
    if (listeners.click) {
      // @ts-ignore
      style['cursor'] = 'pointer'
    }
    const data = { class: `${block} material-icons`, style, on: { ...listeners } }
    let name = props.name
    if (!name) {
      name = 'warning'
      console.warn('[app-icon]', ' ', 'using a fallback icon, please provide a name')
    }

    return h('span', data, name)
  }
}
