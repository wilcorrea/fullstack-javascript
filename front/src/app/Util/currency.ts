/**
 * @param value
 * @param {currency} precision
 * @param {string} thousands
 * @param {string} decimal
 * @param {string} currency
 * @returns {string}
 */
export const money = (
  value: number,
  precision: number = 2,
  thousands: string = '.',
  decimal: string = ',',
  currency: string = 'R$ '
): string => {
  const pieces = Number(value).toFixed(precision).split('.')
  const left: any = pieces[0]
  const right = (pieces.length === 1 ? '' : pieces[1]).padEnd(precision, '0')
  // noinspection RegExpRedundantEscape
  const answer = left
    .reverse()
    .replace(/(\d{3})/g, `$1${thousands}`)
    .reverse()
    .replace(new RegExp(`^\\${thousands}`, 'g'), '')
    .concat(`${decimal}${right}`)
  if (!currency) {
    return answer
  }
  return `${currency}${answer}`
}

/**
 * @param value
 * @param {string} thousands
 * @param {string} decimal
 * @param {string} currency
 * @returns {number | any}
 */
export const unMoney = (
  value: string,
  thousands: string = '.',
  decimal: string = ',',
  currency: string = 'R$ '
): number => {
  // noinspection RegExpRedundantEscape
  const string = String(value)
    .replace(currency, '')
    .replace(new RegExp(`\\${thousands}`, 'g'), '')
    .replace(new RegExp(`\\${decimal}`, 'g'), '.')
  return Number(string)
}
