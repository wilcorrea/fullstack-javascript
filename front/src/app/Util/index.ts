import { get, clone } from './general'

export default {
  get, clone
}
