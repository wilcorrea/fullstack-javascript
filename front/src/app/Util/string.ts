/**
 * @param {string} template
 * @param {Array|Object} replaces
 * @param regex
 * @returns {string}
 */
export const replacement = (
  template: string,
  replaces: any,
  regex?: ((expression: string) => RegExp)
): string => {
  // cast the template to a new string
  const string = template
  // noinspection RegExpRedundantEscape
  let keyFy = (expression: string) => new RegExp(`\\{${expression}\\}`, 'g')
  if (regex) {
    keyFy = regex
  }
  const replace = (replacing: string, key: string, value: string) => {
    return replacing.replace(keyFy(key), value)
  }

  if (Array.isArray(replaces)) {
    return replaces.reduce(
      (replacing, value, index) => replace(replacing, String(index), String(value)),
      string
    )
  }

  if (typeof replaces === 'object') {
    return Object.keys(replaces).reduce(
      (replacing, key) => replace(replacing, key, String(replaces[key])),
      string
    )
  }

  return template
}

/**
 * @param {string} key
 * @returns {RegExp}
 */
export const regexByColon = (key: string) => new RegExp(`:${key}`, 'g')

/**
 * @param {string} content
 * @returns {string}
 */
export const breakLine = (content: string) => content.replace(/ /g, '<br>')
