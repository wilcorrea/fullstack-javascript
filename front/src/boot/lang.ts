import $lang from 'src/lang'

/**
 * @param {Vue} Vue
 * @returns {}
 */
export default (Vue: any) => {
  /**
   */
  Object.defineProperty(Vue.prototype, '$lang', {
    get () {
      return $lang
    }
  })
}
