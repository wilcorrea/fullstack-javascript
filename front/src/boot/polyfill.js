// eslint-disable-next-line no-extend-native
String.prototype.reverse = function () {
  return this.split('').reverse().join('')
}
