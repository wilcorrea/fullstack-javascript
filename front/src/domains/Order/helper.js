/**
 * @returns {Object}
 */
export const statusBackgrounds = () => ({
  WAITING: 'rgb(42, 91, 185)',
  PAID: 'rgb(0, 253, 198)',
  NOT_PAID: 'rgb(255, 0, 0)', // TODO: apply the correct color
  REVERTED: 'rgb(255, 150, 0)' // TODO: apply the correct color
})

/**
 * @returns {Object}
 */
export const statusForegrounds = () => ({
  WAITING: 'rgb(255, 255, 255)',
  PAID: 'rgb(66, 66, 66)',
  NOT_PAID: 'rgb(255, 255, 255)',
  REVERTED: 'rgb(255, 255, 255)'
})
