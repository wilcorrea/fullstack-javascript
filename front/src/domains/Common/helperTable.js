import { get } from 'src/app/Util/general'
import $lang from 'src/lang'
import { breakLine } from 'src/app/Util/string'

/**
 * @param color
 * @returns {Object}
 */
export const style = (color) => ({
  'border-left': `2px solid ${color}`,
  'text-transform': 'uppercase',
  'padding': '0px 0px 0px 5px',
  'font-size': '0.7em',
  'color': 'rgb(60, 73, 93)',
  'line-height': '120%'
})

/**
 * @param {Object} values
 * @param {string} value
 * @param {string} lang
 * @param {boolean} broken
 * @returns {function({h: *, record: *}): *[]}
 */
export const status = (values, value, lang, broken = true) => {
  /**
   * @param h
   * @param record
   * @returns {*}
   */
  return ({ h }) => {
    const answer = (color, content) => {
      const data = {
        domProps: {
          innerHTML: broken ? breakLine(String(content)) : content
        },
        style: style(color)
      }
      return [h('div', data)]
    }
    return answer(get(values, value, {}), $lang(`${lang}.${value}`, value))
  }
}
