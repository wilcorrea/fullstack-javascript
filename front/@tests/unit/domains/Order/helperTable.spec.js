import { mount, shallowMount } from '@vue/test-utils'
import 'src/boot/polyfill.js'

import $lang from 'src/lang'
import { money } from 'src/app/Util/currency'
import { breakLine } from 'src/app/Util/string'
import AppIcon from 'src/app/Components/Icon/AppIcon'

import { statusBackgrounds } from 'src/domains/Order/helper'
import { amount, payments, paymentMethods, status, customer } from 'src/domains/Order/helperTable'

/**
 * @param render
 * @param more
 * @returns {Wrapper<Vue>}
 */
const mounted = (render, more) => mount({
  render (h) {
    return render({ h, ...more })
  },
  ...more
})

/**
 * @param render
 * @param more
 * @returns {Wrapper<Vue>}
 */
const shallowMounted = (render, more) => shallowMount({
  render (h) {
    return h('div', render({ h, ...more }))
  },
  ...more
})

describe('helperTable', () => {
  /**
   * @test
   */
  it('check tableFormat amount', () => {
    const total = Math.random()
    expect(amount({ value: total })).toEqual(money(total))
  })

  /**
   * @test
   */
  it('check tableComponent payments', () => {
    const build = (method, icon) => {
      const record = { payments: [{ fundingInstrument: { method } }] }
      const components = { AppIcon }
      const wrapper = mounted(payments, { record, components })
      return wrapper.html()
    }
    const options = paymentMethods()
    Object.keys(options).forEach((key) => {
      const html = build(key)
      expect(html).toEqual(`<span class="app-icon material-icons">${options[key]}</span>`)
      expect(html).toMatchSnapshot(`payments: ${key}`)
    })
  })

  /**
   * @test
   */
  it('check tableComponent status', () => {
    const build = (value) => {
      const wrapper = mounted(status, { value })
      return wrapper.html()
    }
    const options = statusBackgrounds()
    Object.keys(options).forEach((key) => {
      const html = build(key)
      expect(html).toMatch(`2px solid ${options[key]}`)
      const text = $lang(`domains.order.status.${key}`)
      expect(html).toMatch(breakLine(text))
      expect(html).toMatchSnapshot(`status: ${key}`)
    })
  })

  /**
   * @test
   */
  it('check tableComponent customer', () => {
    const build = (fullname) => {
      const value = { fullname }
      const components = { AppAvatar: { render () { } } }
      const wrapper = shallowMounted(customer, { components, value })
      return wrapper.html()
    }
    const name = 'William Correa'
    const html = build(name)
    expect(html).toMatch(name)
  })
})
