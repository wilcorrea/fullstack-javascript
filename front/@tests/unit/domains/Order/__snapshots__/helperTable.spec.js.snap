// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helperTable check tableComponent payments: payments: BOLETO 1`] = `<span class="app-icon material-icons">receipt</span>`;

exports[`helperTable check tableComponent payments: payments: CREDIT_CARD 1`] = `<span class="app-icon material-icons">credit_card</span>`;

exports[`helperTable check tableComponent payments: payments: DEBIT_CARD 1`] = `<span class="app-icon material-icons">sd_card</span>`;

exports[`helperTable check tableComponent payments: payments: ONLINE_BANK_DEBIT 1`] = `<span class="app-icon material-icons">account_balance_wallet</span>`;

exports[`helperTable check tableComponent payments: payments: ONLINE_BANK_FINANCING 1`] = `<span class="app-icon material-icons">account_balance</span>`;

exports[`helperTable check tableComponent payments: payments: WALLET 1`] = `<span class="app-icon material-icons">attach_money</span>`;

exports[`helperTable check tableComponent status: status: NOT_PAID 1`] = `<div style="border-left: 2px solid rgb(255, 0, 0); text-transform: uppercase; padding: 0px 0px 0px 5px; font-size: 0.7em; color: rgb(60, 73, 93); line-height: 120%;">Pedido<br>Aberto</div>`;

exports[`helperTable check tableComponent status: status: PAID 1`] = `<div style="border-left: 2px solid rgb(0, 253, 198); text-transform: uppercase; padding: 0px 0px 0px 5px; font-size: 0.7em; color: rgb(60, 73, 93); line-height: 120%;">Pedido<br>Pago</div>`;

exports[`helperTable check tableComponent status: status: REVERTED 1`] = `<div style="border-left: 2px solid rgb(255, 150, 0); text-transform: uppercase; padding: 0px 0px 0px 5px; font-size: 0.7em; color: rgb(60, 73, 93); line-height: 120%;">Pedido<br>Estornado</div>`;

exports[`helperTable check tableComponent status: status: WAITING 1`] = `<div style="border-left: 2px solid rgb(42, 91, 185); text-transform: uppercase; padding: 0px 0px 0px 5px; font-size: 0.7em; color: rgb(60, 73, 93); line-height: 120%;">Aguardando<br>Pagamento</div>`;
